src/EventListener/ExceptionListener.php

```diff
@@ -0,0 +1,416 @@
+
+            return;
+        }
+
+        // FIXME: log the exception here if we do not find a better place to log via monolog from outside.
+
+        if (!$this->renderErrorScreens) {
+            return;
+        }
+
+        // Search if any exception in the chain is implementing a http exception.
+        $response = $this->checkHttpExceptions($exception, $event->getRequest());
+        if ($response instanceof Response) {
+            $event->setResponse($this->setXStatusCode($response));
+
+            return;
+        }
+
+        // If still nothing worked out, we wrap it in a plain "internal error occured" message.
+        $event->setResponse($this->renderErrorTemplate('error', 500, $event->getRequest()));
+    }
+
+    /**
+     * Ensure we have a proper X-Status-Code set on a response.
+     *
+     * Setting the status code is needed as otherwise the kernel will not know we handled the exception and will set
+     * an status code of 500 ignoring the status of our response.
+     *
+     * @param Response $response
+     *
+     * @return Response
+     */
+    private function setXStatusCode(Response $response)
+    {
+        if (!$response->headers->has('X-Status-Code')) {
+            $response->headers->set('X-Status-Code', $response->getStatusCode());
+        }
+
+        return $response;
+    }
+
+    /**
+     * Check if the exception or any exception in the chain implements the response exception interface.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @return null|Response
+     */
+    private function checkResponseException($exception)
+    {
+        do {
+            if ($exception instanceof ResponseExceptionInterface) {
+                return $exception->getResponse();
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Check if the exception or any exception in the chain inherits from HttpException and is understood by us.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @param Request $request    The HTTP request.
+     *
+     * @return null|Response
+     */
+    private function checkHttpExceptions(\Exception $exception, $request)
+    {
+        do {
+            if ($response = $this->checkHttpException($exception, $request)) {
+                return $response;
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Check if the Contao 404 should be rendered.
+     *
+     * @param \Exception $exception The exception to test.
+     *
+     * @return bool
+     */
+    private function isRenderingOfContao404PossibleFor($exception)
+    {
+        // If not an 404 exception or the frontend has not been booted, we can not render.
+        if (!($exception instanceof NotFoundHttpException && defined('BE_USER_LOGGED_IN'))) {
```

Comment:
I can not. The main reason this exists is because of the lines the page model like:
https://github.com/contao/core-bundle/blob/52a32d2ee7cea17960e94a509e5f700948552de5/src/Resources/contao/models/PageModel.php#L275
When we end up here, the constant is not necessarily defined and will trigger a warning then.
Therefore it is impossible to use the page handler object (which uses the page model in return).